### PR TITLE
Add pfSense example

### PR DESCRIPTION
**Pull request by b0bcarlson**

### CODE REVIEW

File: example_configs/pfSense.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Configuration for pfSense
+
+Note: Using the test feature in pfSense does not appear to work, and LDAP users do not appear in the Users tab, 
+these are likely bugs in pfSense.
+
+This is only a basic configuration, allowing all valid users in the group to have full permissions.
```
Comment (nitnelave) [inline example_configs/pfSense.md:6]: in the admins group

File: example_configs/pfSense.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+| **Search scope base DN**          | `dc=example,dc=com`                                                                                                                                               |
+| **Authentication containers**     | `ou=people`                                                                                                                                                       |
+| **Extended query**                | enabled                                                                                                                                                           |
+| **Query**                         | `&(objectClass=person)(memberof=cn=admins,ou=groups,dc=example,dc=com)` This will require users to be in the `admins` group. Remove that part to allow all users. |
+| **Bind anonymous**                | disabled (if desired, also set credentials if disabling)                                                                                                          |
```
Comment (nitnelave) [inline example_configs/pfSense.md:26]: LLDAP doesn't support anonymous bind, you can remove the comment.

File: example_configs/pfSense.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+| Group naming attribute            | `cn`                                                                                                                                                              |
+| **Group member attribute**        | `memberUid`                                                                                                                                                       |
+| **RFC 2307 Groups**               | enabled                                                                                                                                                           |
+| RFC 2307 User DN                  | disabled                                                                                                                                                          |
+| **Group Object Class**            | `group`                                                                                                                                                           |
```
Comment (nitnelave) [inline example_configs/pfSense.md:33]: Shouldn't that be `groupOfUniqueNames`?

**Review by nitnelave — CHANGES_REQUESTED** [review]
Thanks for the contribution! Can you also link to this from the Readme?

**Comment by b0bcarlson** [comment]
Something still isn't quite right with this setup. I'm investigating, however the logs from pfSense are a bit lacking

**Review by naziris — COMMENTED** [review]
Does not work quite well.

Group member attribute:  `memberUid`   seems to cause issues. change it that, user is able to authenticate but it does not fetch the group it belong to (?)
Keeping an eye on it though. Thank you for your effort!

File: example_configs/pfSense.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+| **Bind anonymous**                | disabled (if desired, also set credentials if disabling)                                                                                                          |
+| **Bind credentials**              | username: `cn=auth,ou=people,dc=example,dc=com`                                                                                                                   |
+| **User naming attribute**         | `uid`                                                                                                                                                             |
+| Group naming attribute            | `cn`                                                                                                                                                              |
+| **Group member attribute**        | `memberUid`                                                                                                                                                       |
```
Comment (nitnelave) [inline example_configs/pfSense.md:30]: That's just `member`

**Comment by nitnelave** [comment]
Right, `memberUid` should just be `member`

**Comment by naziris** [comment]
> Right, `memberUid` should just be `member`

Changing it to member and Group Object Class to groupOfUniqueNames, requires RFC 2307 Groups to be unchecked in order to authenticate.
Even though when the user log ins, I get the usual [No page assigned to this user! Click here to logout.]

I'll keep investigate, we are close.

**Comment by b0bcarlson** [comment]
I'm not near my computer at the moment so I can't show the logs but there was an issue when I had that second box checked, I believe lldap didn't understand the DN query and returned no groups. I'll pull those logs as soon as I can.

**Comment by nitnelave** [comment]
If that's a blocker, it would be trivial to "implement" `memberUid`

**Comment by b0bcarlson** [comment]
* group member attribute: `memberof`, group object class: `group` (or `groupOfUniqueNames`, same result), and both RFC2307s unchecked:
``` 2023-01-12T17:56:23.723838942+00:00 INFO     LDAP session [ 264ms | 0.08% / 100.00% ]

2023-01-12T17:56:23.723911291+00:00 INFO     ┝━ LDAP request [ 263ms | 0.11% / 99.60% ]

2023-01-12T17:56:23.723928053+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 1, op: BindRequest(LdapBindRequest { dn: "cn=auth,ou=people,dc=bobcodes,dc=net", cred: Simple("********") }), ctrl: [] }

2023-01-12T17:56:23.723931898+00:00 DEBUG    │  ┝━ do_bind [ 263ms | 0.02% / 99.49% ]

2023-01-12T17:56:23.723936218+00:00 DEBUG    │  │  ┝━ 🐛 [debug]: DN: cn=auth,ou=people,dc=bobcodes,dc=net

2023-01-12T17:56:23.723952036+00:00 DEBUG    │  │  ┝━ bind [ 262ms | 0.06% / 99.34% ]

2023-01-12T17:56:23.724392799+00:00 DEBUG    │  │  │  ┕━ passwords_match [ 262ms | 99.28% ]

2023-01-12T17:56:23.986646054+00:00 DEBUG    │  │  ┝━ get_user_groups [ 337µs | 0.13% ]

2023-01-12T17:56:23.986667761+00:00 DEBUG    │  │  │  ┝━ 🐛 [debug]:  | user_id: UserId("auth")

2023-01-12T17:56:23.986753542+00:00 DEBUG    │  │  │  ┝━ 🐛 [debug]:  | query: SELECT "groups"."group_id", "display_name", "creation_date", "uuid" FROM "groups" INNER JOIN "memberships" ON "groups"."group_id" = "memberships"."group_id" WHERE "user_id" = ?

2023-01-12T17:56:23.987315608+00:00 DEBUG    │  │  │  ┕━ 🐛 [debug]:  | return: {GroupDetails { group_id: GroupId(3), display_name: "lldap_strict_readonly", creation_date: 2022-08-01T12:31:09.298822598Z, uuid: Uuid("f2ea4bb9-4840-3262-ac28-dad59dada61f") }}

2023-01-12T17:56:23.987327+00:00 DEBUG    │  │  ┕━ 🐛 [debug]: Success!

2023-01-12T17:56:23.987343758+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: BindResponse(LdapBindResponse { res: LdapResult { code: Success, matcheddn: "", message: "", referral: [] }, saslcreds: None })

2023-01-12T17:56:23.988686131+00:00 INFO     ┝━ LDAP request [ 826µs | 0.10% / 0.31% ]

2023-01-12T17:56:23.988736540+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: SearchRequest(LdapSearchRequest { base: "uid=bobcodes,ou=people,dc=bobcodes,dc=net", scope: Subtree, aliases: InSearching, sizelimit: 0, timelimit: 25, typesonly: false, filter: Equality("uid", "bobcodes"), attrs: ["memberof"] }), ctrl: [] }

2023-01-12T17:56:23.988742205+00:00 DEBUG    │  ┝━ do_search [ 573µs | 0.02% / 0.22% ]

2023-01-12T17:56:23.988769753+00:00 DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "uid=bobcodes,ou=people,dc=bobcodes,dc=net" | scope: User(Equality("uid", "bobcodes"))

2023-01-12T17:56:23.988774058+00:00 DEBUG    │  │  ┕━ get_user_list [ 530µs | 0.02% / 0.20% ]

2023-01-12T17:56:23.988781638+00:00 DEBUG    │  │     ┝━ 🐛 [debug]:  | ldap_filter: And([Equality("uid", "bobcodes"), Equality("uid", "bobcodes")])

2023-01-12T17:56:23.988794963+00:00 DEBUG    │  │     ┝━ 🐛 [debug]:  | parsed_filters: And([UserId(UserId("bobcodes")), UserId(UserId("bobcodes"))])

2023-01-12T17:56:23.988797468+00:00 DEBUG    │  │     ┝━ expand_attribute_wildcards [ 14.7µs | 0.01% ]

2023-01-12T17:56:23.988810293+00:00 DEBUG    │  │     │  ┕━ 🐛 [debug]:  | ldap_attributes: ["memberof"] | resolved_attributes: ["memberof"]

2023-01-12T17:56:23.988819401+00:00 DEBUG    │  │     ┕━ list_users [ 457µs | 0.17% ]

2023-01-12T17:56:23.988828644+00:00 DEBUG    │  │        ┝━ 🐛 [debug]:  | filters: Some(And([UserId(UserId("bobcodes")), UserId(UserId("bobcodes"))])) | get_groups: true

2023-01-12T17:56:23.989012425+00:00 DEBUG    │  │        ┝━ 🐛 [debug]:  | query: SELECT "users"."user_id", "email", "users"."display_name", "first_name", "last_name", "avatar", "users"."creation_date", "users"."uuid", "groups"."group_id", "groups"."display_name" AS "group_display_name", "groups"."creation_date" AS "group_creation_date", "groups"."uuid" AS "group_uuid" FROM "users" LEFT JOIN "memberships" ON "users"."user_id" = "memberships"."user_id" LEFT JOIN "groups" ON "memberships"."group_id" = "groups"."group_id" WHERE "users"."user_id" = ? AND "users"."user_id" = ? ORDER BY "users"."user_id" ASC, "group_display_name" ASC

2023-01-12T17:56:23.989701506+00:00 DEBUG    │  │        ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: UserId("bobcodes"), email: "bob@bobcodes.net", display_name: "bobcodes", first_name: "Bob", last_name: "Carlson", avatar: JpegPhoto([]), creation_date: 2022-08-01T12:32:25.325253469Z, uuid: Uuid("7ee87955-b92b-30fd-a750-0117a75ad1dd") }, groups: Some([GroupDetails { group_id: GroupId(4), display_name: "admins", creation_date: 2022-08-18T17:07:10.803078137Z, uuid: Uuid("dad4c18e-232d-3fe5-bf88-a43cb3074c8b") }, GroupDetails { group_id: GroupId(1), display_name: "lldap_admin", creation_date: 2022-08-01T12:31:09.215534862Z, uuid: Uuid("34152503-7830-3713-a8fe-17d855433e34") }, GroupDetails { group_id: GroupId(5), display_name: "pfsense_admin", creation_date: 2022-09-16T18:17:04.894310854Z, uuid: Uuid("d60126f7-73e4-3027-8ffc-b3f11fd6c4aa") }]) }]

2023-01-12T17:56:23.989766647+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | response: SearchResultEntry(LdapSearchResultEntry { dn: "uid=bobcodes,ou=people,dc=bobcodes,dc=net", attributes: [LdapPartialAttribute { atype: "memberof", vals: [[117, 105, 100, 61, 97, 100, 109, 105, 110, 115, 44, 111, 117, 61, 103, 114, 111, 117, 112, 115, 44, 100, 99, 61, 98, 111, 98, 99, 111, 100, 101, 115, 44, 100, 99, 61, 110, 101, 116], [117, 105, 100, 61, 108, 108, 100, 97, 112, 95, 97, 100, 109, 105, 110, 44, 111, 117, 61, 103, 114, 111, 117, 112, 115, 44, 100, 99, 61, 98, 111, 98, 99, 111, 100, 101, 115, 44, 100, 99, 61, 110, 101, 116], [117, 105, 100, 61, 112, 102, 115, 101, 110, 115, 101, 95, 97, 100, 109, 105, 110, 44, 111, 117, 61, 103, 114, 111, 117, 112, 115, 44, 100, 99, 61, 98, 111, 98, 99, 111, 100, 101, 115, 44, 100, 99, 61, 110, 101, 116]] }] })

2023-01-12T17:56:23.989897377+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: SearchResultDone(LdapResult { code: Success, matcheddn: "", message: "", referral: [] })

2023-01-12T17:56:23.991572019+00:00 INFO     ┕━ LDAP request [ 14.9µs | 0.01% ]

2023-01-12T17:56:23.991583412+00:00 DEBUG       ┕━ 🐛 [debug]:  | msg: LdapMsg { msgid: 3, op: UnbindRequest, ctrl: [] }

2023-01-12T17:56:42.927844054+00:00 INFO     LDAP session [ 81.5µs | 100.00% ]
```

This all looks correct to me, but pfSense returns auth successful with no groups.

* Same as above but first RFC2307 (LDAP Server uses RFC 2307 style group membership) checked:
```2023-01-12T18:00:42.334937082+00:00 INFO     ┝━ LDAP request [ 1.12ms | 0.07% / 0.24% ]

2023-01-12T18:00:42.334964458+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: SearchRequest(LdapSearchRequest { base: "ou=people,dc=bobcodes,dc=net", scope: Subtree, aliases: InSearching, sizelimit: 0, timelimit: 25, typesonly: false, filter: Equality("uid", "bobcodes"), attrs: [] }), ctrl: [] }

2023-01-12T18:00:42.334969928+00:00 DEBUG    │  ┝━ do_search [ 769µs | 0.02% / 0.16% ]

2023-01-12T18:00:42.335030423+00:00 DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "ou=people,dc=bobcodes,dc=net" | scope: Users

2023-01-12T18:00:42.335045326+00:00 DEBUG    │  │  ┕━ get_user_list [ 682µs | 0.04% / 0.15% ]

2023-01-12T18:00:42.335062867+00:00 DEBUG    │  │     ┝━ 🐛 [debug]:  | ldap_filter: Equality("uid", "bobcodes")

2023-01-12T18:00:42.335084104+00:00 DEBUG    │  │     ┝━ 🐛 [debug]:  | parsed_filters: UserId(UserId("bobcodes"))

2023-01-12T18:00:42.335087550+00:00 DEBUG    │  │     ┝━ expand_attribute_wildcards [ 55.6µs | 0.01% ]

2023-01-12T18:00:42.335139878+00:00 DEBUG    │  │     │  ┕━ 🐛 [debug]:  | ldap_attributes: [] | resolved_attributes: ["objectclass", "uid", "mail", "givenname", "sn", "cn", "jpegPhoto", "createtimestamp"]

2023-01-12T18:00:42.335168091+00:00 DEBUG    │  │     ┕━ list_users [ 444µs | 0.10% ]

2023-01-12T18:00:42.335188899+00:00 DEBUG    │  │        ┝━ 🐛 [debug]:  | filters: Some(UserId(UserId("bobcodes"))) | get_groups: false

2023-01-12T18:00:42.335342441+00:00 DEBUG    │  │        ┝━ 🐛 [debug]:  | query: SELECT "users"."user_id", "email", "users"."display_name", "first_name", "last_name", "avatar", "users"."creation_date", "users"."uuid" FROM "users" WHERE "users"."user_id" = ? ORDER BY "users"."user_id" ASC

2023-01-12T18:00:42.335958574+00:00 DEBUG    │  │        ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: UserId("bobcodes"), email: "bob@bobcodes.net", display_name: "bobcodes", first_name: "Bob", last_name: "Carlson", avatar: JpegPhoto([]), creation_date: 2022-08-01T12:32:25.325253469Z, uuid: Uuid("7ee87955-b92b-30fd-a750-0117a75ad1dd") }, groups: None }]

2023-01-12T18:00:42.336147819+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | response: SearchResultEntry(LdapSearchResultEntry { dn: "uid=bobcodes,ou=people,dc=bobcodes,dc=net", attributes: [LdapPartialAttribute { atype: "objectclass", vals: [[105, 110, 101, 116, 79, 114, 103, 80, 101, 114, 115, 111, 110], [112, 111, 115, 105, 120, 65, 99, 99, 111, 117, 110, 116], [109, 97, 105, 108, 65, 99, 99, 111, 117, 110, 116], [112, 101, 114, 115, 111, 110]] }, LdapPartialAttribute { atype: "uid", vals: [[98, 111, 98, 99, 111, 100, 101, 115]] }, LdapPartialAttribute { atype: "mail", vals: [[98, 111, 98, 64, 98, 111, 98, 99, 111, 100, 101, 115, 46, 110, 101, 116]] }, LdapPartialAttribute { atype: "givenname", vals: [[66, 111, 98]] }, LdapPartialAttribute { atype: "sn", vals: [[67, 97, 114, 108, 115, 111, 110]] }, LdapPartialAttribute { atype: "cn", vals: [[98, 111, 98, 99, 111, 100, 101, 115]] }, LdapPartialAttribute { atype: "createtimestamp", vals: [[50, 48, 50, 50, 45, 48, 56, 45, 48, 49, 84, 49, 50, 58, 51, 50, 58, 50, 53, 46, 51, 50, 53, 50, 53, 51, 52, 54, 57, 43, 48, 48, 58, 48, 48]] }] })

2023-01-12T18:00:42.336357819+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: SearchResultDone(LdapResult { code: Success, matcheddn: "", message: "", referral: [] })

2023-01-12T18:00:42.337472138+00:00 INFO     ┝━ LDAP request [ 345µs | 0.05% / 0.07% ]

2023-01-12T18:00:42.337520031+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 3, op: SearchRequest(LdapSearchRequest { base: "ou=people,dc=bobcodes,dc=net", scope: Subtree, aliases: InSearching, sizelimit: 0, timelimit: 25, typesonly: false, filter: And([Equality("memberof", "bobcodes"), And([Equality("objectClass", "person"), Equality("memberof", "cn=pfsense_admin,ou=groups,dc=bobcodes,dc=net")])]), attrs: [] }), ctrl: [] }

2023-01-12T18:00:42.337527958+00:00 DEBUG    │  ┝━ do_search [ 122µs | 0.01% / 0.03% ]

2023-01-12T18:00:42.337551652+00:00 DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "ou=people,dc=bobcodes,dc=net" | scope: Users

2023-01-12T18:00:42.337555296+00:00 DEBUG    │  │  ┕━ get_user_list [ 87.9µs | 0.02% ]

2023-01-12T18:00:42.337569922+00:00 DEBUG    │  │     ┕━ 🐛 [debug]:  | ldap_filter: And([Equality("memberof", "bobcodes"), And([Equality("objectClass", "person"), Equality("memberof", "cn=pfsense_admin,ou=groups,dc=bobcodes,dc=net")])])

2023-01-12T18:00:42.337684428+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: SearchResultDone(LdapResult { code: UnwillingToPerform, matcheddn: "", message: "Unsupported user filter: while parsing a group ID: Missing DN value", referral: [] })

2023-01-12T18:00:42.338940760+00:00 INFO     ┕━ LDAP request [ 17.3µs | 0.00% ]

2023-01-12T18:00:42.338954086+00:00 DEBUG       ┕━ 🐛 [debug]:  | msg: LdapMsg { msgid: 4, op: UnbindRequest, ctrl: [] }

2023-01-12T18:00:42.949551972+00:00 INFO     LDAP session [ 71.6µs | 100.00% ]
```
pfSense returns auth failed

 * Same as above but with the second box checked as well (RFC 2307 Use DN for username search.)
```2023-01-12T18:04:35.025303624+00:00 INFO     LDAP session [ 441ms | 0.09% / 100.00% ]

2023-01-12T18:04:35.025465562+00:00 INFO     ┝━ LDAP request [ 440ms | 0.09% / 99.62% ]

2023-01-12T18:04:35.025513417+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 1, op: BindRequest(LdapBindRequest { dn: "cn=auth,ou=people,dc=bobcodes,dc=net", cred: Simple("********") }), ctrl: [] }

2023-01-12T18:04:35.025531137+00:00 DEBUG    │  ┝━ do_bind [ 439ms | 0.03% / 99.53% ]

2023-01-12T18:04:35.025548603+00:00 DEBUG    │  │  ┝━ 🐛 [debug]: DN: cn=auth,ou=people,dc=bobcodes,dc=net

2023-01-12T18:04:35.025591538+00:00 DEBUG    │  │  ┝━ bind [ 439ms | 0.08% / 99.38% ]

2023-01-12T18:04:35.026707298+00:00 DEBUG    │  │  │  ┕━ passwords_match [ 438ms | 99.30% ]

2023-01-12T18:04:35.465024479+00:00 DEBUG    │  │  ┝━ get_user_groups [ 546µs | 0.12% ]

2023-01-12T18:04:35.465070880+00:00 DEBUG    │  │  │  ┝━ 🐛 [debug]:  | user_id: UserId("auth")

2023-01-12T18:04:35.465204475+00:00 DEBUG    │  │  │  ┝━ 🐛 [debug]:  | query: SELECT "groups"."group_id", "display_name", "creation_date", "uuid" FROM "groups" INNER JOIN "memberships" ON "groups"."group_id" = "memberships"."group_id" WHERE "user_id" = ?

2023-01-12T18:04:35.466239937+00:00 DEBUG    │  │  │  ┕━ 🐛 [debug]:  | return: {GroupDetails { group_id: GroupId(3), display_name: "lldap_strict_readonly", creation_date: 2022-08-01T12:31:09.298822598Z, uuid: Uuid("f2ea4bb9-4840-3262-ac28-dad59dada61f") }}

2023-01-12T18:04:35.466257173+00:00 DEBUG    │  │  ┕━ 🐛 [debug]: Success!

2023-01-12T18:04:35.466295181+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: BindResponse(LdapBindResponse { res: LdapResult { code: Success, matcheddn: "", message: "", referral: [] }, saslcreds: None })

2023-01-12T18:04:35.467799887+00:00 INFO     ┝━ LDAP request [ 953µs | 0.07% / 0.22% ]

2023-01-12T18:04:35.467826322+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: SearchRequest(LdapSearchRequest { base: "ou=people,dc=bobcodes,dc=net", scope: Subtree, aliases: InSearching, sizelimit: 0, timelimit: 25, typesonly: false, filter: Equality("uid", "bobcodes"), attrs: [] }), ctrl: [] }

2023-01-12T18:04:35.467831485+00:00 DEBUG    │  ┝━ do_search [ 643µs | 0.02% / 0.15% ]

2023-01-12T18:04:35.467890989+00:00 DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "ou=people,dc=bobcodes,dc=net" | scope: Users

2023-01-12T18:04:35.467895789+00:00 DEBUG    │  │  ┕━ get_user_list [ 544µs | 0.03% / 0.12% ]

2023-01-12T18:04:35.467901996+00:00 DEBUG    │  │     ┝━ 🐛 [debug]:  | ldap_filter: Equality("uid", "bobcodes")

2023-01-12T18:04:35.467911903+00:00 DEBUG    │  │     ┝━ 🐛 [debug]:  | parsed_filters: UserId(UserId("bobcodes"))

2023-01-12T18:04:35.467928373+00:00 DEBUG    │  │     ┝━ expand_attribute_wildcards [ 31.5µs | 0.01% ]

2023-01-12T18:04:35.467957387+00:00 DEBUG    │  │     │  ┕━ 🐛 [debug]:  | ldap_attributes: [] | resolved_attributes: ["objectclass", "uid", "mail", "givenname", "sn", "cn", "jpegPhoto", "createtimestamp"]

2023-01-12T18:04:35.467972050+00:00 DEBUG    │  │     ┕━ list_users [ 392µs | 0.09% ]

2023-01-12T18:04:35.467980177+00:00 DEBUG    │  │        ┝━ 🐛 [debug]:  | filters: Some(UserId(UserId("bobcodes"))) | get_groups: false

2023-01-12T18:04:35.468107352+00:00 DEBUG    │  │        ┝━ 🐛 [debug]:  | query: SELECT "users"."user_id", "email", "users"."display_name", "first_name", "last_name", "avatar", "users"."creation_date", "users"."uuid" FROM "users" WHERE "users"."user_id" = ? ORDER BY "users"."user_id" ASC

2023-01-12T18:04:35.468864090+00:00 DEBUG    │  │        ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: UserId("bobcodes"), email: "bob@bobcodes.net", display_name: "bobcodes", first_name: "Bob", last_name: "Carlson", avatar: JpegPhoto([]), creation_date: 2022-08-01T12:32:25.325253469Z, uuid: Uuid("7ee87955-b92b-30fd-a750-0117a75ad1dd") }, groups: None }]

2023-01-12T18:04:35.469020276+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | response: SearchResultEntry(LdapSearchResultEntry { dn: "uid=bobcodes,ou=people,dc=bobcodes,dc=net", attributes: [LdapPartialAttribute { atype: "objectclass", vals: [[105, 110, 101, 116, 79, 114, 103, 80, 101, 114, 115, 111, 110], [112, 111, 115, 105, 120, 65, 99, 99, 111, 117, 110, 116], [109, 97, 105, 108, 65, 99, 99, 111, 117, 110, 116], [112, 101, 114, 115, 111, 110]] }, LdapPartialAttribute { atype: "uid", vals: [[98, 111, 98, 99, 111, 100, 101, 115]] }, LdapPartialAttribute { atype: "mail", vals: [[98, 111, 98, 64, 98, 111, 98, 99, 111, 100, 101, 115, 46, 110, 101, 116]] }, LdapPartialAttribute { atype: "givenname", vals: [[66, 111, 98]] }, LdapPartialAttribute { atype: "sn", vals: [[67, 97, 114, 108, 115, 111, 110]] }, LdapPartialAttribute { atype: "cn", vals: [[98, 111, 98, 99, 111, 100, 101, 115]] }, LdapPartialAttribute { atype: "createtimestamp", vals: [[50, 48, 50, 50, 45, 48, 56, 45, 48, 49, 84, 49, 50, 58, 51, 50, 58, 50, 53, 46, 51, 50, 53, 50, 53, 51, 52, 54, 57, 43, 48, 48, 58, 48, 48]] }] })

2023-01-12T18:04:35.469210773+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: SearchResultDone(LdapResult { code: Success, matcheddn: "", message: "", referral: [] })

2023-01-12T18:04:35.470844473+00:00 INFO     ┝━ LDAP request [ 309µs | 0.05% / 0.07% ]

2023-01-12T18:04:35.470880628+00:00 DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 3, op: SearchRequest(LdapSearchRequest { base: "ou=people,dc=bobcodes,dc=net", scope: Subtree, aliases: InSearching, sizelimit: 0, timelimit: 25, typesonly: false, filter: And([Equality("memberof", "uid=bobcodes,ou=people,dc=bobcodes,dc=net"), And([Equality("objectClass", "person"), Equality("memberof", "cn=pfsense_admin,ou=groups,dc=bobcodes,dc=net")])]), attrs: [] }), ctrl: [] }

2023-01-12T18:04:35.470885684+00:00 DEBUG    │  ┝━ do_search [ 102µs | 0.01% / 0.02% ]

2023-01-12T18:04:35.470904384+00:00 DEBUG    │  │  ┝━ 🐛 [debug]:  | request.base: "ou=people,dc=bobcodes,dc=net" | scope: Users

2023-01-12T18:04:35.470907204+00:00 DEBUG    │  │  ┕━ get_user_list [ 72.6µs | 0.02% ]

2023-01-12T18:04:35.470923621+00:00 DEBUG    │  │     ┕━ 🐛 [debug]:  | ldap_filter: And([Equality("memberof", "uid=bobcodes,ou=people,dc=bobcodes,dc=net"), And([Equality("objectClass", "person"), Equality("memberof", "cn=pfsense_admin,ou=groups,dc=bobcodes,dc=net")])])

2023-01-12T18:04:35.471015667+00:00 DEBUG    │  ┕━ 🐛 [debug]:  | response: SearchResultDone(LdapResult { code: UnwillingToPerform, matcheddn: "", message: "Unsupported user filter: Unexpected group DN format. Got \"uid=bobcodes,ou=people,dc=bobcodes,dc=net\", expected: \"cn=groupname,ou=groups,dc=bobcodes,dc=net\"", referral: [] })

2023-01-12T18:04:35.472193111+00:00 INFO     ┕━ LDAP request [ 17.6µs | 0.00% ]

2023-01-12T18:04:35.472206468+00:00 DEBUG       ┕━ 🐛 [debug]:  | msg: LdapMsg { msgid: 4, op: UnbindRequest, ctrl: [] }
```
pfSense again reports auth failed.

So yes, I would say I agree that we are close. The first test seems to work correctly but like the group object class is incorrect so it can't find the groups?

**Comment by nitnelave** [comment]
A couple of things from your logs: I think the first one should work, it's surprising that pfsense doesn't like it.
The second one has `memberOf=bobcodes` instead of `memberOf=cn=bobcodes,...` which is not supported by LLDAP. Is that from RFC2307?
The last one has `memberOf=bobcodes,ou=people,...` <- trying to search for a group in `ou=people`, that's not gonna work.

**Comment by naziris** [comment]
I haven't figure out anything yet, but I will place here my current state with the options I have set at the moment just for reference.

```
Search scope:
	Level: Entire Subtree
	Base DN: dc=domain,dc=ext

Authentication containers: ou=people

Extended query: Enabled
Query: &(objectClass=person)(memberOf=cn=vpn_users,ou=groups,dc=domain,dc=ext)

Bind credentials: uid=admin,ou=people,dc=domain,dc=ext     MyVeryStr0ngPassw0rd

User naming attribute: uid
Group naming attribute: cn
Group member attribute: memberOf

RFC 2307 Groups: Disabled

Group Object Class: group
```

The error I receive while logging in:

```
2023-01-18T08:17:10.299410814+00:00 INFO     LDAP session [ 77.8ms | 0.23% / 100.00% ]
2023-01-18T08:17:10.299510441+00:00 INFO     ┝━ LDAP request [ 77.0ms | 99.05% ]
2023-01-18T08:17:10.377174202+00:00 INFO     ┕━ LDAP request [ 558µs | 0.72% ]
2023-01-18T08:17:10.378069437+00:00 WARN        ┕━ 🚧 [warn]: Ignoring unknown group attribute ""memberof"" in filter.\n\
                                To disable this warning, add it to "ignored_group_attributes" in the config.
```

The memberof attribute that is reported in the error log is the one in the Query field. Despite of that, the user is able to login but no groups are coming along.

**Comment by b0bcarlson** [comment]
Ok, I decided to give this another shot. I went with what naziris used, but did `member` instead of `memberOf` which resulted in the `Ignoring unrecognized group attribute` error (this should read user attribute, not group; I'll make a PR for this fix momentarily). I also switched to using `display_name` instead of `cn` for group name (is this what it's supposed to be?)

I noticed that it dumps out the groups I'm part of, though. Here it is when I used `memberOf` (`member` does the same thing, but gives the attribute error).
https://gist.github.com/b0bcarlson/b4e24c9cf662ec0bd6d8a12e18a47f6c
Line 75 and again on line 105.

I guess what I'm most curious about though is lines 137 and 139. This is the very end, so it's where I expected to see the groups but 137 dumps out a bunch of integers which I don't understand and 139 has an empty result.

**Comment by lordratner** [comment]
> I haven't figure out anything yet, but I will place here my current state with the options I have set at the moment just for reference.
> 
> ```
> Search scope:
> 	Level: Entire Subtree
> 	Base DN: dc=domain,dc=ext
> 
> Authentication containers: ou=people
> 
> Extended query: Enabled
> Query: &(objectClass=person)(memberOf=cn=vpn_users,ou=groups,dc=domain,dc=ext)
> 
> Bind credentials: uid=admin,ou=people,dc=domain,dc=ext     MyVeryStr0ngPassw0rd
> 
> User naming attribute: uid
> Group naming attribute: cn
> Group member attribute: memberOf
> 
> RFC 2307 Groups: Disabled
> 
> Group Object Class: group
> ```
> 
> The error I receive while logging in:
> 
> ```
> 2023-01-18T08:17:10.299410814+00:00 INFO     LDAP session [ 77.8ms | 0.23% / 100.00% ]
> 2023-01-18T08:17:10.299510441+00:00 INFO     ┝━ LDAP request [ 77.0ms | 99.05% ]
> 2023-01-18T08:17:10.377174202+00:00 INFO     ┕━ LDAP request [ 558µs | 0.72% ]
> 2023-01-18T08:17:10.378069437+00:00 WARN        ┕━ 🚧 [warn]: Ignoring unknown group attribute ""memberof"" in filter.\n\
>                                 To disable this warning, add it to "ignored_group_attributes" in the config.
> ```
> 
> The memberof attribute that is reported in the error log is the one in the Query field. Despite of that, the user is able to login but no groups are coming along.

Have you created the group in pfsense and added permissions?

From the documentation: https://docs.netgate.com/pfsense/en/latest/usermanager/ldap.html#ldap-groups

"There are two requirements for LDAP groups to function properly:

The LDAP authentication settings must match the group membership style used by the LDAP server

The same groups must exist locally ([Manage Local Groups](https://docs.netgate.com/pfsense/en/latest/usermanager/groups.html))"

**Comment by lordratner** [comment]
This is now working for me.

**Pull request closed without merging**